### PR TITLE
HDDS-3947: Sort DNs for client when the key is a file for #getFileStatus #listStatus APIs

### PR DESCRIPTION
**Pull request by rakeshadr**
## What changes were proposed in this pull request?

Similar to OzoneManagerFS#lookupFile(OmKeyArgs args, String clientAddress) interface, "clientAddress" has been passed as an argument to OzoneManagerFS#listStatus(), OzoneManagerFS#getFileStatus() APIs.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-3947

## How was this patch tested?

Used existing TestKeyManagerImpl unit test cases.

### CODE REVIEW

File: hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/rpc/RpcClient.java [hunk 1]
```diff
@@ -1016,6 +1016,7 @@ public OzoneFileStatus getOzoneFileStatus(String volumeName,
         .setBucketName(bucketName)
         .setKeyName(keyName)
         .setRefreshPipeline(true)
+        .setSortDatanodesInPipeline(topologyAwareReadEnabled)
```
Comment (bshashikant) [inline hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/rpc/RpcClient.java:1019]: @rakeshadr , can you plz add some details here why sorting of dns is required here?

Comment (rakeshadr) [reply hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/rpc/RpcClient.java:1019]: Thanks @bshashikant for the reviews. This change is basically to make "getFileStatus" and "listStatus" apis similar to "lookupFile()" API, where it does sorting.

Reference:
https://github.com/apache/hadoop-ozone/pull/1164#discussion_r451122204

**Review by adoroszlai — COMMENTED** [review]
Thanks @rakeshadr for working on this.  Can you please update the patch to resolve conflicts?

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/fs/OzoneManagerFS.java [hunk 2]
```diff
@@ -31,7 +31,18 @@
...
-  OzoneFileStatus getFileStatus(OmKeyArgs args) throws IOException;
+
+  /**
+   * Get file status for a file or a directory.
+   *
+   * @param args          the args of the key provided by client.
+   * @param clientAddress a hint to key manager, order the datanode in returned
+   *                      pipeline by distance between client and datanode.
+   * @return file status.
+   * @throws IOException
```
[the remark this answers is not in the document]
Comment (rakeshadr) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/fs/OzoneManagerFS.java:42]: No specific reason, I will describe exception cases. 
I will update the KeyManagerImpl implementation as well.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/fs/OzoneManagerFS.java [hunk 3]
```diff
@@ -31,7 +31,18 @@
...
+   */
+  OzoneFileStatus getFileStatus(OmKeyArgs args, String clientAddress)
```
[the remark this answers is not in the document]
Comment (rakeshadr) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/fs/OzoneManagerFS.java:44]: Agreed, will keep the existing #getFileStatus(args) method.

**Review by rakeshadr — COMMENTED** [review]
Thanks @adoroszlai for the reviews

**Comment by elek** [comment]
I am afraid something wrong happened with the last rebase/merge:  

```
Files changed 737
```

**Comment by rakeshadr** [comment]
@bshashikant , @adoroszlai I have created fresh PR https://github.com/apache/hadoop-ozone/pull/1385 as this PR has been polluted:-)

I have addressed review comments in that new PR.

**Pull request closed without merging**